File: pkg/core/event.go
package core

import "github.com/shopspring/decimal"

// On-chain events

// NodeBalanceUpdatedEvent represents the NodeBalanceUpdated event
type NodeBalanceUpdatedEvent struct {
	BlockchainID uint64          `json:"blockchain_id"`
	Asset        string          `json:"asset"`
	Balance      decimal.Decimal `json:"balance"`
}

// HomeChannelCreatedEvent represents the ChannelCreated event
type HomeChannelCreatedEvent channelEvent

// HomeChannelMigratedEvent represents the ChannelMigrated event
type HomeChannelMigratedEvent channelEvent

// HomeChannelCheckpointedEvent represents the Checkpointed event
type HomeChannelCheckpointedEvent channelEvent

// HomeChannelChallengedEvent represents the Challenged event
type HomeChannelChallengedEvent channelChallengedEvent

// HomeChannelClosedEvent represents the Closed event
type HomeChannelClosedEvent channelEvent

// EscrowDepositInitiatedEvent represents the EscrowDepositInitiated event
type EscrowDepositInitiatedEvent channelEvent

// EscrowDepositChallengedEvent represents the EscrowDepositChallenged event
type EscrowDepositChallengedEvent channelChallengedEvent

// EscrowDepositFinalizedEvent represents the EscrowDepositFinalized event
type EscrowDepositFinalizedEvent channelEvent

// EscrowDepositsPurgedEvent represents the EscrowDepositsPurged event emitted when expired
// escrow deposits are finalized by the purge queue without a signed FINALIZE_ESCROW_DEPOSIT state.
type EscrowDepositsPurgedEvent struct {
	// EscrowIDs holds the hex-encoded escrow IDs (== channel_id in the channels table) that were purged.
	EscrowIDs []string `json:"escrow_ids"`
}

// EscrowWithdrawalInitiatedEvent represents the EscrowWithdrawalInitiated event
type EscrowWithdrawalInitiatedEvent channelEvent

// EscrowWithdrawalChallengedEvent represents the EscrowWithdrawalChallenged event
type EscrowWithdrawalChallengedEvent channelChallengedEvent

// EscrowWithdrawalFinalizedEvent represents the EscrowWithdrawalFinalized event
type EscrowWithdrawalFinalizedEvent channelEvent

type channelEvent struct {
	ChannelID    string `json:"channel_id"`
	StateVersion uint64 `json:"state_version"`
	// UserSig is the hex-encoded user signature recovered from the on-chain state payload.
	// Empty when the parsed event carries no user signature (e.g. unilateral node-only state).
	UserSig string `json:"user_sig,omitempty"`
}

type channelChallengedEvent struct {
	ChannelID       string `json:"channel_id"`
	StateVersion    uint64 `json:"state_version"`
	ChallengeExpiry uint64 `json:"challenge_expiry"`
	// UserSig is the hex-encoded user signature recovered from the on-chain state payload.
	UserSig string `json:"user_sig,omitempty"`
}

// ValidatorRegisteredEvent is emitted by ChannelHub when the node registers a new
// signature validator. Users should react to unexpected registrations by revoking
// ERC20 approvals granted to ChannelHub — see contracts/SECURITY.md for details.
type ValidatorRegisteredEvent struct {
	BlockchainID uint64 `json:"blockchain_id"`
	ValidatorID  uint8  `json:"validator_id"`
	// Validator is the EIP-55 checksummed hex address of the registered validator contract.
	// Always compare using strings.EqualFold or common.HexToAddress(ev.Validator).Hex()
	// to avoid silent mismatches against lowercase or non-checksummed config values.
	Validator   string `json:"validator"`
	BlockNumber uint64 `json:"block_number"` // block where the event was emitted; use as fromBlock on reconnect
}

type BlockchainEvent struct {
	ContractAddress string `json:"contract_address"`
	BlockchainID    uint64 `json:"blockchain_id"`
	Name            string `json:"name"`
	BlockNumber     uint64 `json:"block_number"`
	TransactionHash string `json:"transaction_hash"`
	LogIndex        uint32 `json:"log_index"`
}

File: sdk/ts/src/core/event.ts
import { Address, Hex } from 'viem';

// ============================================================================
// Base Event Types
// ============================================================================

/**
 * Base channel event structure
 */
export interface ChannelEvent {
  channelId: string;
  stateVersion: bigint; // uint64
}

/**
 * Channel challenged event structure (includes challenge expiry)
 */
export interface ChannelChallengedEvent {
  channelId: string;
  stateVersion: bigint; // uint64
  challengeExpiry: bigint; // uint64
}

/**
 * Blockchain event metadata
 */
export interface BlockchainEvent {
  contractAddress: Address;
  blockchainId: bigint; // uint64
  name: string;
  blockNumber: bigint; // uint64
  transactionHash: Hex;
  logIndex: number; // uint32
}

// ============================================================================
// Home Channel Events
// ============================================================================

/**
 * HomeChannelCreatedEvent represents the ChannelCreated event
 */
export type HomeChannelCreatedEvent = ChannelEvent;

/**
 * HomeChannelMigratedEvent represents the ChannelMigrated event
 */
export type HomeChannelMigratedEvent = ChannelEvent;

/**
 * HomeChannelCheckpointedEvent represents the Checkpointed event
 */
export type HomeChannelCheckpointedEvent = ChannelEvent;

/**
 * HomeChannelChallengedEvent represents the Challenged event
 */
export type HomeChannelChallengedEvent = ChannelChallengedEvent;

/**
 * HomeChannelClosedEvent represents the Closed event
 */
export type HomeChannelClosedEvent = ChannelEvent;

// ============================================================================
// Escrow Deposit Events
// ============================================================================

/**
 * EscrowDepositInitiatedEvent represents the EscrowDepositInitiated event
 */
export type EscrowDepositInitiatedEvent = ChannelEvent;

/**
 * EscrowDepositChallengedEvent represents the EscrowDepositChallenged event
 */
export type EscrowDepositChallengedEvent = ChannelChallengedEvent;

/**
 * EscrowDepositFinalizedEvent represents the EscrowDepositFinalized event
 */
export type EscrowDepositFinalizedEvent = ChannelEvent;

// ============================================================================
// Escrow Withdrawal Events
// ============================================================================

/**
 * EscrowWithdrawalInitiatedEvent represents the EscrowWithdrawalInitiated event
 */
export type EscrowWithdrawalInitiatedEvent = ChannelEvent;

/**
 * EscrowWithdrawalChallengedEvent represents the EscrowWithdrawalChallenged event
 */
export type EscrowWithdrawalChallengedEvent = ChannelChallengedEvent;

/**
 * EscrowWithdrawalFinalizedEvent represents the EscrowWithdrawalFinalized event
 */
export type EscrowWithdrawalFinalizedEvent = ChannelEvent;

// ============================================================================
// Validator Events
// ============================================================================

/**
 * Emitted when the node registers a new signature validator on ChannelHub.
 * Users must react to unexpected registrations by revoking ERC20 approvals
 * granted to ChannelHub — see contracts/SECURITY.md.
 *
 * `validator` is always EIP-55 checksummed. Compare with getAddress(ev.validator)
 * to avoid silent mismatches against lowercase or non-checksummed config values.
 *
 * Pass `blockNumber + 1n` as `fromBlock` on reconnect for gap-free monitoring.
 */
export interface ValidatorRegisteredEvent {
    blockchainId: bigint;
    validatorId: number; // uint8
    validator: Address; // EIP-55 checksummed
    blockNumber: bigint; // use as fromBlock on reconnect
}
